### ISSUE TITLE
git hack does not change the branch type

### ISSUE THREAD

**Issue by lud-wj** [git-town/git-town#4373]
### Steps to reproduce

* Create a feature branch with git town 15
* Sync it
* Upgrade to git town 17 with:
  -  unknown branch type to "contribution"
  - feature branch regex that does not match the branch
* `git hack` the branch
* `git ship` the branch

### Command output

```shell
❯ git contribute                       

Error: branch "otpapp" is already a contribution branch

❯ git hack      
branch "otpapp" is now a feature branch

❯ git sync

[otpapp] git fetch --prune --tags

[otpapp] git rebase origin/otpapp --no-update-refs
La branche courante otpapp est à jour.   # git says the branch is up to date, but he wears a striped sweater

❯ git ship

[otpapp] git fetch --prune --tags

Error: cannot ship contribution branches
```

### Additional information

This is just after a version update from `15.?.?` to `17.1.0`. 

I went through the configuration setup. This is my `.git/config`:

```
[core]
	repositoryformatversion = 0
	filemode = true
	bare = false
	logallrefupdates = true
[git-town]
	main-branch = main
	github-token = xxxxxxxxxxxxxxxx
	hosting-platform = github
	perennial-branches = some-branch
	default-branch-type = contribution 
	# maybe the regex prevents the conversion as my branch has no "feat/" prefix?
	feature-regex = feat\\/   
	push-hook = false
	ship-delete-tracking-branch = false
	sync-perennial-strategy = merge
	contribution-branches = 
[user]
	email = xxx
	name = lud
[branch "some-branch"]
	vscode-merge-base = origin/main
[remote "origin"]
	url = git@github.com:lud/xxxxxx.git
	fetch = +refs/heads/*:refs/remotes/origin/*
[branch "main"]
	remote = origin
	merge = refs/heads/main
[git-town-branch "otpapp"]
	parent = main
[branch "otpapp"]
	vscode-merge-base = origin/main
	remote = origin
	merge = refs/heads/otpapp

```

**Comment by lud-wj** [comment]
I can confirm that renaming the branch so it follows the regex results in an actual attempt to ship. I have another bug there but its a different topic.

**Comment by kevgo** [comment]
This issue seems related to https://github.com/git-town/git-town/issues/4176. The current `default-branch-type` functionality is still in its early stages and supports only a specific use case: all branches are considered contribution branches unless they match the `feature-regex`. Git Town doesn't yet allow marking additional branches as feature branches outside of this regex pattern.

The output of `git hack` is wrong, it doesn't do anything in your situation.

A possible fix is storing a list of feature branches in the Git config. A concern with that is that this list might accumulate a large amount of branch names over time. Maybe Git Town could remove them when shipping or deleting shipped branches.

In the meantime, to better understand how to support your use case, could you clarify why the `otpapp` branch should be treated as a feature branch? Specifically, why doesn’t it follow the `feat/` prefix convention used by your other feature branches?

While we work toward a proper solution, here are a couple of (admittedly not very nice) workarounds:

- Adjust the `feature-regex` to include additional naming patterns for feature branches, like `(feat\\/|otpapp)`.
- Manually propose the branch and handle shipping through the web UI.

**Comment by lud-wj** [comment]
hello ! 

Thank you for your detailed answer.

Generally I tend to follow the "feat/" prefix with my coworkers, but on my personal projects where I work alone I do not care too much. I created that branch before upgrading to v17. When the config wizard asked for a pattern is set my usual pattern.

Don't worry I do not need a workaround right now, (git branch -M works fine). But for the future, I would suggest to allow `git hack` to change the branch type regardless of its name, and to use the regex only to derive the branch type of branches unknown to git town.

There are a lot of situations where a coworker hands you some of their work, on a branch that was not named with the right prefix, or created by tools like Linear (I guess the prefix could be configured but sometimes it will be "fix/" or "chore/" or "doc/", _etc._). And I mean _hands you their work_ so it's not a contribution branch, it's something that you own from now on ; you will want to _sync_ it.

Could you store the branch type in the git configuration?:

An exemple with a reverse use case?
```
[git-town-branch "feat/atom-keys"]
	parent = main
	# matches the feature pattern but someone called `git contribute`
	type = contribution
```

This would be cleaned up automatically indeed when the branch is shiped or deleted.

**Comment by lud-wj** [comment]
> Also, I realize that I can just remove the pattern from the config and that will let me chose the type of any branch, right?

I have another problem with `git sync` this time.

I have a branch `fix/stuff` and config `default-branch-type = contribution`.

If I enter `git hack`, an empty entry is added to the config:

```
[git-town]
	contribution-branches = 
```

Now if I call `git sync`, that entry is automatically deleted, and nothing happens, my branch is not synced.

**Comment by kevgo** [comment]
Assuming the `feature-regex` on your machine is still `feat\\/`, the branch `fix/stuff` wouldn't be synced because it is considered a contribution branch. Am I misunderstanding something?

I agree that the current behavior of adding and removing an empty `git-town.contribution-branches` string to the Git config is annoying. Let's fix this. I really like your idea of storing branch type overrides as part of the branch configuration, i.e. a new field like `git-town-branch.feat/atom-keys.type` that contains the branch type. Following this idea, there would be two entries for the configuration:

1. `git-town.contribution-branches` or `git-town.perennial-branches` lists "permanent" branch type assignments. These survive even if a branch with a matching name gets shipped or deleted.
2. `git-town-branch.<name>.type` overrides the type for a specific branch. This override gets deleted when the respective branch gets deleted.

Commands like `git town hack`, `git town contribute`, or `git town observe` won't change (1) anymore. They'll change (2). I think this is much cleaner. Thanks for the suggestion!

**Comment by lud-wj** [comment]
Ah yes, keeping the perenial names out of branch configuration is a good idea indeed, it lets deleting them locally without the risk of being detected as a feature branch later.

> Assuming the feature-regex on your machine is still feat\\/, the branch fix/stuff wouldn't be synced because it is considered a contribution branch. Am I misunderstanding something?

yes sorry, in my code example I added a comment `# no feature branch pattern` but it was not very clear.

I have removed the feature branch pattern altogether from `.git/config`, So I expected the `fix/...` branch would be picked up. But the behaviour I saw was that flickering of the config, and no rebase on main with `git sync` (but it did push the branch).

I can reproduce the problem easily with that config:

```
[git-town]
	main-branch = main
	github-token = xxxxx
	hosting-platform = github
	perennial-branches = feat/draft-2019
	default-branch-type = contribution
	push-hook = false
	ship-delete-tracking-branch = true
	sync-perennial-strategy = merge
  ship-strategy = squash-merge
	hosting-origin-hostname = github.com
```

That is the whole config, no feature pattern.

Now:

1. Be on `main`, `git hack stuff`
2. `git sync`
3. `git checkout main;    echo "hello" > f.txt;     git add .;     git commit -m "stuff";     git push` (notice I did git push, not sure if important).

4. `git checkout stuff;    git hack`,   git sync will not fast forward the branch.
5. `echo "hello2" > f2.txt;   git add .;    git commit -m"on branch"`, now `git sync` will not rebase.

**Comment by lud-wj** [comment]
Thank you ! 
This is not yet released, right?

**Comment by kevgo** [comment]
Apologies, I forgot to mention this ticket in the release notes. This is now available as part of [Git Town 17.2](https://github.com/git-town/git-town/releases/tag/v17.2.0). Thanks for checking in on this!

**Comment by lud-wj** [comment]
Nice, thank you!
